students/kegan/session07/html_render.py

```diff
@@ -0,0 +1,186 @@
+        """
+        file_out.write(self.get_tag(ind))
+        file_out.write('\n')
+        for c in self.content:
+            try:
+                c.render(file_out, ind + 1)
+            except AttributeError:
+                file_out.write('{}{}\n'.format((ind + 1) * self.indent, c))
+        file_out.write(self.get_tag(ind, style='close'))
+        file_out.write('\n')
+
+    def get_tag(self, ind, style='open'):
+        """ Generates an HTML tag according to given style
+        with given level of indent. Style can be 'open',
+        'close', or 'selfclosing' (defaults to 'open').
+        Args:
+            ind (str) : level of indent
+            style (str) : style to use
+        Returns:
+            tag : formatted tag
+        """
+        substrings = {
+            'indent': ind * self.indent,
+            'tag': self.tag,
+            'attributes':
+                '' if style == 'close' else ''.join([
+                    ' {}="{}"'.format(key, self.attrs[key])
+                    for key in sorted(self.attrs)]),
+            'close': '/' if style == 'close' else '',
+            'selfclosing': ' /' if style == 'selfclosing' else ''}
+        tag = '{indent}<{close}{tag}{attributes}{selfclosing}>'
+        tag = tag.format(**substrings)
+        return tag
+
+
+class OneLineTag(Element):
+    """ Handles ne-line HTML elements. """
+
+    def render(self, file_out, ind=0):
+        """ Renders the contents of this Element into HTML format
+        on one line and writes to file using given file handler.
+        Args:
+            file_out (TextIOWrapper) : file handler
+            ind (str) : level of indentation for this Element
+        """
+        file_out.write(self.get_tag(ind))
+        for c in self.content:
+            file_out.write(c)
+        file_out.write(self.get_tag(0, style='close'))
+        file_out.write('\n')
+
+
+class SelfClosingTag(Element):
+    """ Handles contentless HTML elements. """
+
+    def render(self, file_out, ind=0):
+        """ Renders this contentless Element into HTML
+        and writes to file using given file handler.
+        Args:
+            file_out (TextIOWrapper) : file handler
+            ind (str) : level of in
+        """
+        file_out.write(self.get_tag(ind, 'selfclosing'))
+        file_out.write('\n')
+
+
+class Html(Element):
+    """ HTML page. """
+    tag = 'html'
+
+    def render(self, file_out, ind=0):
+        """ Renders the contents of this Element into HTML
+        format headed by doctype. Writes to file using given file handler.
+        Args:
+            file_out (TextIOWrapper) : file handler
+            ind (str) : level of indentation for this Element
+        """
+        file_out.write('<!DOCTYPE {}>\n'.format(self.tag))
+        Element.render(self, file_out, ind=1)
+
+
+class A(OneLineTag):
+    """ HTML hyperlink element. """
+    tag = 'a'
+
+    def __init__(self, link, content):
```

Comment:
I actually looked it up -- anchor tags can take other attributes as well. So this should take attribures as kwargs and add the href one to them.

But the right approach, for sure.